src/Libraries/Microsoft.Extensions.AI.Abstractions/Functions/AIFunctionArguments.cs

```diff
@@ -20,36 +21,42 @@ namespace Microsoft.Extensions.AI;
 /// container.
 /// </remarks>
-public sealed class AIFunctionArguments : IDictionary<string, object?>, IReadOnlyDictionary<string, object?>
+public class AIFunctionArguments : IDictionary<string, object?>, IReadOnlyDictionary<string, object?>
 {
     /// <summary>The nominal arguments.</summary>
     private readonly Dictionary<string, object?> _arguments;
 
     /// <summary>Initializes a new instance of the <see cref="AIFunctionArguments"/> class.</summary>
-    public AIFunctionArguments()
+    /// <param name="comparer">The optional <see cref="IEqualityComparer{T}"/> to use for key comparisons.</param>
+    public AIFunctionArguments(IEqualityComparer<string>? comparer = null)
     {
-        _arguments = [];
+        _arguments = new Dictionary<string, object?>(comparer);
     }
 
     /// <summary>
     /// Initializes a new instance of the <see cref="AIFunctionArguments"/> class containing
     /// the specified <paramref name="arguments"/>.
     /// </summary>
     /// <param name="arguments">The arguments represented by this instance.</param>
+    /// <param name="comparer">The <see cref="IEqualityComparer{T}"/> to be used.</param>
     /// <remarks>
     /// The <paramref name="arguments"/> reference will be stored if the instance is
-    /// already a <see cref="Dictionary{TKey, TValue}"/>, in which case all dictionary
-    /// operations on this instance will be routed directly to that instance. If <paramref name="arguments"/>
-    /// is not a dictionary, a shallow clone of its data will be used to populate this
-    /// instance. A <see langword="null"/> <paramref name="arguments"/> is treated as an
-    /// empty dictionary.
+    /// already a <see cref="Dictionary{TKey, TValue}"/> and no <paramref name="comparer"/> is specified,
+    /// in which case all dictionary operations on this instance will be routed directly to that instance.
+    /// If <paramref name="arguments"/> is not a dictionary or a <paramref name="comparer"/> is specified,
+    /// a shallow clone of its data will be used to populate this instance.
+    /// A <see langword="null"/> <paramref name="arguments"/> is treated as an empty dictionary.
     /// </remarks>
-    public AIFunctionArguments(IDictionary<string, object?>? arguments)
+    public AIFunctionArguments(IDictionary<string, object?>? arguments, IEqualityComparer<string>? comparer = null)
     {
-        _arguments =
-            arguments is null ? [] :
-            arguments as Dictionary<string, object?> ??
-            new Dictionary<string, object?>(arguments);
+        _arguments = comparer is null
+            ? arguments is null
+                ? []
+                : arguments as Dictionary<string, object?> ??
+                  new Dictionary<string, object?>(arguments)
+            : arguments is null
+                ? new Dictionary<string, object?>(comparer)
+                : new Dictionary<string, object?>(arguments, comparer);
```

Comment:
What if the comparer matches the one already used by the dictionary?

Reply:
I don't have access to the comparer in the `IDictionary` interface to be able to compare or am I missing something here?